### PR TITLE
Fix more C warnings

### PR DESCRIPTION
**Pull request by tamird**
Revival of #628 (sorry, it fell off my radar while you were investigating the statement issues).

I think this should be good to merge.

### CODE REVIEW

**Comment by sodabrew** [comment]
Needs a rebase since the rubocop PR included the gperf changes, and other comments.

File: ext/mysql2/extconf.rb [hunk 1]
```diff
@@ -94,18 +94,32 @@ def asplode lib
...
-  try_link('int main() {return 0;}', flag)
-end.each do |flag|
+# -Wall and -Wextra are included by default.
+[
+  '-fsanitize=address',
+  '-fsanitize=cfi',
+  '-fsanitize=integer',
+  '-fsanitize=memory',
+  '-fsanitize=thread',
+  '-fsanitize=undefined',
```
Comment (sodabrew) [inline ext/mysql2/extconf.rb]: I don't think people would want to go to production with these?

File: ext/mysql2/extconf.rb [hunk 2]
```diff
@@ -94,18 +94,32 @@ def asplode lib
...
+  '-Werror',
+  '-Weverything',
```
Comment (sodabrew) [inline ext/mysql2/extconf.rb]: I'm concerned that -Weverything is an open-ended list, so future compilers will add in more checks that we might not pass. I would use -Weverything without -Werror, but given the number of allowances below, I don't want to use -Weverything.

Comment (tamird) [inline ext/mysql2/extconf.rb]: We could use `-Werror` only in CI?

**Comment by sodabrew** [comment]
I really like the idea of -Werror only in CI. Would help in cases like #674

**Comment by tamird** [comment]
hurray!

**Pull request merged**